changes/2024-TBD_cache-across-hierarchy-keyrings/change.md

```diff
@@ -0,0 +1,552 @@
+        )
+  - [Decrypt](https://github.com/aws/aws-encryption-sdk-python/blob/1a1213a31776477dcc4aab44b2a4dc2eb514113e/src/aws_encryption_sdk/caches/__init__.py#L101): as per [spec](https://github.com/awslabs/aws-encryption-sdk-specification/tree/ffb2b0cc6a956b2cec3a33be3c3672605b6907fb/framework/caching-cmm.md#appendix-a-cache-entry-identifier-formulas)
+    - EDK_HASHES = [SHA512(SerializeEncryptedDataKey(key)) for key in decryptMaterialsRequest.encryptedDataKeys]
+    - ENTRY_ID = SHA512(
+      SHA512(UTF8Encode(cachingCMM.partitionId)) + AlgorithmSuiteId(decryptMaterialsRequest.algorithmSuite) + CONCATENATE(SORTED(EDK_HASHES)) + PADDING_OF_512_ZERO_BITS + SHA512(SerializeEncryptionContext(decryptMaterialsRequest.encryptionContext))
+      )
+- The length of cache identifier for EncryptionMaterials
+  and DecryptionMaterials is 64 bytes.
+
+### DB-ESDK Searchable Encryption Cache Identifier formula
+
+- To fetch BeaconKeyMaterials from the
+  CMC for Searchable Encryption,
+  the cache identifier for a beacon key is a key-id. The key-id is a
+  [string](https://github.com/aws/aws-database-encryption-sdk-dynamodb/blob/b5705ee12257fb18f867478bf17ba31f50c26c8b/DynamoDbEncryption/dafny/DynamoDbEncryption/src/SearchInfo.dfy#L185),
+  and here is an example initialization
+  [[ref](https://github.com/aws/aws-database-encryption-sdk-dynamodb/blob/b5705ee12257fb18f867478bf17ba31f50c26c8b/TestVectors/dafny/DDBEncryption/src/JsonConfig.dfy#L527)].
+- The length of cache identifier for BeaconKeyMaterials is “variable” based on the key-id.
+- For single-tenant, the customer sets the key-id directly in the config. For multi-tenant, they set an attribute name, and the key comes out of that attribute of the item.
+
+## Birthday Bound for SHA512 and SHA384
+
+A [birthday attack](https://en.wikipedia.org/wiki/Birthday_attack)
+is a bruteforce collision attack that exploits the
+mathematics behind the birthday problem in probability
+theory. This attack can be used to abuse communication
+between two or more parties. The attack depends on the
+higher likelihood of collisions found between random
+attack attempts and a fixed degree of
+permutations (pigeonholes). With a birthday attack,
+if we have a total of n bits output and therefore
+$2^n$ different outputs, it is possible to find a
+collision of a hash function with 50% chance by
+calculating approximately $sqrt(2^n) = 2^{n/2}$ hashes.
+
+To find a random collision in a SHA512 hash with
+probability 0.5, we would need to calculate
+approximately $2^256$ hashes. Similarly for SHA384,
+we would need to calculate $2^192$ hashes.
+Therefore, SHA512 is more collision resistant.
+
+## Appendix
+
+## A1: Relating Cache Entries (Branch Keys) to KMS Relationships
+
+For some MPL Consumers,
+Cache Entries SHOULD only be served IF
+the request is being served under the same KMS Relationship,
+which is the most secure condition,
+as it prevents a Confused Deputy scenario
+(detailed in **Confused Deputy 1**).
```

Comment:
Issue/Suggestion: Remove.
I never wrote this; it was on my TODO list.
Alternatively, if you want to write, you can. 
But I do not think it is needed today.
```suggestion
(detailed in **Confused Deputy 1**).
```

Reply:
I'll remove this as there are more pertinent things to focus on.